core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java

```diff
@@ -211,33 +212,34 @@ public int getTopologyId() {
 
    @Override
    public void addBackupLockForKey(Object key) {
-      // we need to synchronize this collection to be able to get a valid snapshot from another thread during state transfer
-      if (backupKeyLocks == null) backupKeyLocks = Collections.synchronizedSet(new HashSet<>(INITIAL_LOCK_CAPACITY));
-      backupKeyLocks.add(key);
+      // we need a synchronized collection to be able to get a valid snapshot from another thread during state transfer
+      final Set<Object> keys = backupKeyLocks.updateAndGet((value) -> value == null ? Collections.synchronizedSet(new HashSet<>(INITIAL_LOCK_CAPACITY)) : value);
```

Comment:
No need for synchronization between writers, we already have a `TransactionSynchronizerInterceptor` dealing with that.